### ISSUE TITLE
Template location: ABSPATH vs WP_CONTENT_DIR

### ISSUE THREAD

**Issue by enricodeleo** [wpovernight/woocommerce-pdf-invoices-packing-slips#14]
Hi!
Deploying Wordpress with tools like Trellis or Bedrock the path to the custom invoice template changes and it is switched back to simple.
It seems that the path in the db is saved as the real path and not the path exposed to the web server (eg the webserver exposes /srv/www/site.ext/current that is a symlink to the current deployed version of the app like /srv/www/site.ext/releases/timestamp). Other plugins just use /current without any problems...

Any idea?

**Comment by Spreeuw** [comment]
This is most likely an old installation that has the absolute path to the template stored. Newer versions (after 1.4.4)  store the link to the template relative to  ABSPATH, e.g. starting with wp-content/. You can fix this by resaving the template settings before deploying.

**Comment by enricodeleo** [comment]
Thanks for your reply!
Actually, the version I'm using is 1.5.26. In settings, the plugin says that the path is `/srv/www/domain.com/releases/20160125180536/web/app/themes/mytheme/woocommerce/pdf/yourtemplate`

**Comment by Spreeuw** [comment]
What I meant is that the setting is from 1.4.4 (even when you're on 1.5.26).
If you resave the template settings, it should replace that old version setting (an absolute path) with the new relative path.

**Comment by enricodeleo** [comment]
I'm saved again but nothing changed, plus I have to re-save on every deployment because of the issue above.

These are the relevant lines about ABSPATH and paths in general in my config (boilerplate from [bedrock](https://github.com/roots/bedrock)):

```php
$root_dir = dirname(__DIR__);
$webroot_dir = $root_dir . '/web';
define('CONTENT_DIR', '/app');
define('WP_CONTENT_DIR', $webroot_dir . CONTENT_DIR);
define('WP_CONTENT_URL', WP_HOME . CONTENT_DIR);
```

**Comment by Spreeuw** [comment]
That's really strange!
As for the re-save, I actually mean that you'd do that once. Once you're on relative paths, it will never go back to absolute paths anymore. But ultimately, it does need ABSPATH. I looked in the WP source, but for as far as I can see, ABSPATH is not relying on any site specific configurations. I have never used bedrock though, so this may be different. I do see that their folder structure is different.

Can you confirm that the plugin is storing relative paths (wp-content/ etc) in the settings?
You can see the code that collects these paths [here](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/master/includes/class-wcpdf-settings.php#L1309-L1343). It uses a combination of get_stylesheet_directory(), get_template_directory() and ABSPATH.

**Comment by enricodeleo** [comment]
I'll play with ABSPATH with this implementation and keep you posted. I think that looking forward to a better (and more solid) approach to wordpress as a framework/cmf and approaches like bedrock need to be encouraged.

**Comment by Spreeuw** [comment]
Ok, thanks! Definitely let me know what you find!
Just a side remark - you wrote 'Other plugins just use /current without any problems', but this plugin is different in the sense that it works with paths rather than urls, and needs to store the path in the DB (rather than calling it directly because there can be multiple PDF templates installed).

**Comment by batruji** [comment]
@enricodeleo did you manage to find a solution to this problem? I am using bedrock and experiencing the same situation where I have to re-save this setting after every deploy. 

Option for pdf template still appears as absolute path, using version 1.5.24:
/var/www/mysite/releases/20160128162925/site/web/app/themes/mytheme/woocommerce/pdf/mycustomtemplate

@Spreeuw is there a way to hardcode this string as a relative path in the database? If yes what should it be?

Thank you.

**Comment by enricodeleo** [comment]
@batruji in config/application.php the application's root directory is gathered with

`$root_dir = dirname(__DIR__);`

This line gets the real path of the directory that is the one within releases (named with a timestamp). If you are into hardcoding stuffs you could change that line into

`$root_dir = '/src/www/example.com/current';`

This will change every path variable/constant, ABSPATH included (that is the constant used with this plugin).

**Comment by stewarty** [comment]
I have this issue as well.  

I've changed the path in the database to use our Symlinked path but it changes every time we do a deploy.

Stewart

**Comment by batruji** [comment]
@stewarty this is how I solved the issue. In application.php:

```
$root_dir = dirname(__DIR__);

/**
 * fix for PDF plugin path on server update
 */

// find where /releases/ string is in root dir path
$release_string_position = strpos ( $root_dir, '/releases/' );
if ( $release_string_position !== FALSE ) {
  // change root dir to point at 'current' release
  $root_dir = substr_replace($root_dir, '/current', $release_string_position , 24);
}

$webroot_dir = $root_dir . '/web';
```
The solution is not optimal but works ;)

**Comment by Spreeuw** [comment]
I think I have solved this issue, see 8c41dbc. I need to maintain backwards compatibility to the `ABSPATH` setting, but I have made a workaround that checks if `WP_CONTENT_DIR` is within the `ABSPATH` folder, and if not, it uses `WP_CONTENT_DIR` to determine the basepath. I have tested this with a local bedrock installation, please let me know if this indeed solves your issue @stewarty, @batruji, @enricodeleo!

**Comment by Spreeuw** [comment]
I released 2.0.8 today which includes the aforementioned commit. I'm closing this issue for now - let me know if any problems with Bedrock still occur!
